spanner/session.go

```diff
@@ -582,6 +610,27 @@ func newSessionPool(sc *sessionClient, config SessionPoolConfig) (*sessionPool,
 	if config.healthCheckSampleInterval == 0 {
 		config.healthCheckSampleInterval = time.Minute
 	}
+	if config.CloseInactiveTransactions || config.LogInactiveTransactions {
+		config.TrackSessionHandles = true
```

Comment:
@olavloite 
For the `CloseInactiveTransactions` or `LogInactiveTransactions` functionality we need to keep a track of every session that is checkout from the pool. Existing `TrackSessionHandles` property does this explicitly where it tracks every session in the list in [newSessionHandle](https://github.com/googleapis/google-cloud-go/blob/2b99e4f39be20fe21e8bc5c1ec1c0e758222c46e/spanner/session.go#L757) method. That is why it has to be made `true` when this feature is rolled out.
Otherwise we would have to write similar logic separately for this feature, which seems redundant. With out tracking sessions we cannot identify long running sessions for removal.

Please help me out if I am missing something here?

Reply:
I think that you can re-use much of the logic that is used for that, but I also think that it would make sense to make sure that the [capturing of the stack trace](https://github.com/googleapis/google-cloud-go/blob/2b99e4f39be20fe21e8bc5c1ec1c0e758222c46e/spanner/session.go#L761) of the go routine that checks out the session is still optional and turned off by default.

So what I would suggest would be to:
1. Keep the `TrackSessionHandles` property false by default as it is at the moment.
2. Change [this if condition](https://github.com/googleapis/google-cloud-go/blob/2b99e4f39be20fe21e8bc5c1ec1c0e758222c46e/spanner/session.go#L757) into something like `if p.TrackSessionHandles || config.CloseInactiveTransactions || config.LogInactiveTransactions`
3. Add an extra `if p.TrackSessionHandles` block around [this line](https://github.com/googleapis/google-cloud-go/blob/2b99e4f39be20fe21e8bc5c1ec1c0e758222c46e/spanner/session.go#L761).
4. Make sure that all code that uses the tracked session handles is able to handle the fact that the stack trace could be nil.

The reason for the above is that the `debug.Stack()` call is a bit expensive, and it's a debug call that you do not necessarily want to get called for everything that checks a session out of the pool (especially as it is turned off by default at the moment).

Reply:
Thanks @olavloite.
I have made changes to enable stack trace explicitly only when `TrackSessionHandles` is enabled for long running sessions detection.